File: app/src/main/java/projects/csce/evence/di/appscope/ViewModelModule.java
package projects.csce.evence.di.appscope;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import dagger.Binds;
import dagger.Module;
import dagger.multibindings.IntoMap;
import daniel.perez.core.di.ViewModelFactory;
import daniel.perez.generateqrview.GenerateQrViewModel;
import projects.csce.evence.di.viewmodel.ViewModelKey;
import projects.csce.evence.viewmodel.MainViewModel;
import daniel.perez.qrcameraview.viewmodel.QrReaderViewModel;

@Module
public abstract class ViewModelModule
{
    @Binds
    abstract ViewModelProvider.Factory bindViewModelFactory(ViewModelFactory factory);

    @Binds
    @IntoMap
    @ViewModelKey(MainViewModel.class)
    abstract ViewModel mainViewModel(MainViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(GenerateQrViewModel.class)
    abstract ViewModel generateQrViewModel(GenerateQrViewModel viewModel);

    @Binds
    @IntoMap
    @ViewModelKey(QrReaderViewModel.class)
    abstract ViewModel qrReaderViewModel(QrReaderViewModel viewModel);
}

File: app/src/main/java/projects/csce/evence/di/appscope/ViewModelModule.kt
package projects.csce.evence.di.appscope

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import dagger.Binds
import dagger.Module
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import dagger.multibindings.IntoMap
import daniel.perez.core.di.ViewModelFactory
import daniel.perez.generateqrview.GenerateQrViewModel
import daniel.perez.qrcameraview.viewmodel.QrReaderViewModel
import projects.csce.evence.di.viewmodel.ViewModelKey
import projects.csce.evence.viewmodel.MainViewModel

//@Module
//@InstallIn(SingletonComponent::class)
abstract class ViewModelModule {
//    @Binds
    abstract fun bindViewModelFactory(factory: ViewModelFactory?): ViewModelProvider.Factory?

//    @Binds
//    @IntoMap
//    @ViewModelKey(GenerateQrViewModel::class)
    abstract fun generateQrViewModel(viewModel: GenerateQrViewModel?): ViewModel?

//    @Binds
//    @IntoMap
//    @ViewModelKey(QrReaderViewModel::class)
    abstract fun qrReaderViewModel(viewModel: QrReaderViewModel?): ViewModel?
}